### PR TITLE
Rust

### LINKED ISSUE #801: Please add Rust grammar to this repo
I saw this ticket: #323 that the other repo is outdated, I searched a bit and I only found this: https://doc.rust-lang.org/grammar.html Maybe a talk to the Mozilla gues could be helpful too?

So all in all, would be great to see rust here.

Regards

Chris

### PR DESCRIPTION
**Pull request by studentmain**
major difference with official parser: 
- `>>` is a parser rule, in order to parse `A<B<C>>` correctly and reject `let a = b > > c` correctly
- identifier is parser rule too, in ANTLR, `async` can't be KW_ASYNC and IDENTIFIER at same time
- macro fragment specifier check is not implemented, implement it will add some contextual keyword, which is too expensive for this purpose
- limitation about isolated \r is not implemented due to the rarity of isolated \r and added complexity in lexer rule.
- All experimental feature are not implemented, as I said in comment, I copy paste from official reference, and they're not here

close #801

### CODE REVIEW

File: rust/rustLexer.g4 [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+ The above copyright notice and this permission notice (including the next paragraph) shall be included in all copies or
+ substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+ WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+lexer grammar rustLexer
```
Comment (KvanTTT) [inline rust/rustLexer.g4:18]: We use upper case for naming, please rename:

```suggestion
lexer grammar RustLexer
```

File: rust/rustParser.g4 [hunk 2]
```diff
@@ -0,0 +1,714 @@
...
+ The above copyright notice and this permission notice (including the next paragraph) shall be included in all copies or
+ substantial portions of the Software.
+
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+ WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+parser grammar rustParser
```
Comment (KvanTTT) [inline rust/rustParser.g4:18]:
```suggestion
parser grammar RustParser
```

File: rust/rustParser.g4 [hunk 3]
```diff
@@ -0,0 +1,714 @@
...
+   | '=>'
+   | '#'
+   //| '$' | '?'
+   ;
+
+// LA can be removed, legal rust code still pass but the cost is `let c = a < < b` will pass... i hope antlr5 can add
+// some new syntax? dsl? for these stuff so i needn't write it in (at least) 5 language
+
+shl: '<' {_input.LA(1)=='<'}? '<';
+shr: '>' {_input.LA(1)=='>'}? '>';
```
Comment (KvanTTT) [inline rust/rustParser.g4:714]: Maybe extract these predicates into base class as we made for other grammars (JavaScript, PHP)?

Comment (studentmain) [reply rust/rustParser.g4:714]: I'll do it after test and other stuff ready.

File: rust/.gitignore [hunk 4]
```diff
@@ -0,0 +1,5 @@
+*.java
```
Comment (KvanTTT) [inline rust/.gitignore:1]: What's the purpose of adding *.java? These files are being generated automatically?

File: rust/RustParser.g4 [hunk 5]
```diff
@@ -0,0 +1,726 @@
...
+parser grammar RustParser
+   ;
+
+options
+{
+   tokenVocab = RustLexer;
+   superClass = RustParserBase;
+}
+// entry point
+crate: innerAttribute* item* EOF;
```
Comment (KvanTTT) [inline rust/RustParser.g4:27]: It's better to unify the formatting of rules:

```suggestion
crate
    : innerAttribute* item* EOF
    ;
```

File: rust/.gitignore (hunk 4 above)
Comment (studentmain) [reply rust/.gitignore:1]: Most java file are generated by `antlr *.g4`, base classes are not included by next line `!*Base.java`

File: rust/RustParser.g4 [hunk 6]
```diff
@@ -0,0 +1,726 @@
...
+module: 'mod' identifier ';' | 'mod' identifier '{' innerAttribute* item* '}';
+
+externCrate: 'extern' 'crate' crateRef asClause? ';';
+crateRef: identifier | 'self';
+asClause: 'as' (identifier | '_');
+
+useDeclaration: 'use' useTree ';';
+useTree
+   : (simplePath? '::')? '*'
+   | (simplePath? '::')? '{' ( useTree (',' useTree)* ','?)? '}'
```
Comment (KvanTTT) [inline rust/RustParser.g4:146]: Merge alternatives?

```suggestion
   : (simplePath? '::')? ('*' | '{' ( useTree (',' useTree)* ','?)? '}')
```

File: rust/RustParser.g4 [hunk 7]
```diff
@@ -0,0 +1,726 @@
...
+
+implementation: inherentImpl | traitImpl;
+inherentImpl
+   : 'impl' generics? type whereClause? '{' innerAttribute* inherentImplItem* '}'
+   ;
+inherentImplItem
+   : outerAttribute*
+   (
+      macroInvocationSemi
+      | ( visibility? ( constantItem | function | method))
```
Comment (KvanTTT) [inline rust/RustParser.g4:231]: Useless parenthesis:

```suggestion
      | visibility? (constantItem | function | method)
```

File: rust/RustParser.g4 [hunk 8]
```diff
@@ -0,0 +1,726 @@
...
+   )
+   ;
+traitImpl
+   : 'unsafe'? 'impl' generics? '!'? typePath 'for' type whereClause? '{' innerAttribute* traitImplItem* '}'
+   ;
+traitImplItem
+   : outerAttribute*
+   (
+      macroInvocationSemi
+      | (visibility? ( typeAlias | constantItem | method | function))
```
Comment (KvanTTT) [inline rust/RustParser.g4:241]: Useless parentheses:

```suggestion
      | visibility? (typeAlias | constantItem | method | function)
```

File: rust/RustParser.g4 [hunk 9]
```diff
@@ -0,0 +1,726 @@
...
+   )
+   ;
+
+externBlock: 'extern' abi? '{' innerAttribute* externalItem* '}';
+externalItem
+   : outerAttribute*
+   (
+      macroInvocationSemi
+      | ( visibility? ( externalStaticItem | externalFunctionItem))
```
Comment (KvanTTT) [inline rust/RustParser.g4:250]:
```suggestion
      | visibility? (externalStaticItem | externalFunctionItem)
```

**Review by KvanTTT — COMMENTED** [review]
It's good in general, but it's better to use the same formatting for all rules:

```antlr
rule
    : alternative1
    | alternative2
    ;
```

**Comment by studentmain** [comment]
FYI, I'm using vscode ANTLR plugin with following config
```json
{
    "antlr4.format": {
        "alignTrailingComments": false,
        "allowShortBlocksOnASingleLine": true,
        "breakBeforeBraces": true,
        "columnLimit": 120,
        "continuationIndentWidth": 3,
        "indentWidth": 3,
        "keepEmptyLinesAtTheStartOfBlocks": true,
        "maxEmptyLinesToKeep": 1,
        "reflowComments": true,
        "spaceBeforeAssignmentOperators": true,
        "tabWidth": 3,
        "useTab": false,
        "alignColons": "hanging",
        "singleLineOverrulesHangingColon": true,
        "allowShortRulesOnASingleLine": false,
        "alignSemicolons": "hanging",
        "breakBeforeParens": true,
        "ruleInternalsOnSingleLine": false,
        "minEmptyLines": 0,
        "groupedAlignments": true,
        "alignFirstTokens": false,
        "alignLexerCommands": false,
        "alignActions": false,
        "alignLabels": true,
        "alignTrailers": false
    }
}
```

**Pull request merged**